src/freertos_drivers/esp32/Esp32BootloaderHal.hxx

```diff
@@ -0,0 +1,562 @@
+    esp_ota_handle_t ota_handle;
+
+    /// Internal flag to indicate that we have initialized the TWAI peripheral and
+    /// should deinit it before exit.
+    bool twai_initialized;
+
+    /// GPIO pin connected to the CAN transceiver TX pin.
+    gpio_num_t tx_pin;
+
+    /// GPIO pin connected to the CAN transceiver RX pin.
+    gpio_num_t rx_pin;
+};
+
+/// Bootloader configuration data.
+static Esp32BootloaderState esp_bl_state;
+
+/// Maximum time to wait for a TWAI frame to be received or transmitted before
+/// giving up.
+static constexpr BaseType_t MAX_TWAI_WAIT = pdMS_TO_TICKS(250);
+
+/// Flag used to indicate that we have been requested to enter the bootloader
+/// instead of normal node operations. Note that this value will not be
+/// initialized by the system and a check for power on reset will need to be
+/// made to initialize it on first boot.
+static uint32_t RTC_NOINIT_ATTR bootloader_request;
+
+/// Value to be assigned to @ref bootloader_request when the bootloader should
+/// run instead of normal node operations.
+static constexpr uint32_t RTC_BOOL_TRUE = 1;
+
+/// Default value to assign to @ref bootloader_request when the ESP32-C3 starts
+/// the first time or when the bootloader should not be run.
+static constexpr uint32_t RTC_BOOL_FALSE = 0;
+
+extern "C"
+{
+
+/// Callback from the bootloader to configure and start the TWAI hardware.
+void bootloader_hw_init(void)
+{
+    // TWAI driver timing configuration, 125kbps.
+    twai_timing_config_t t_config = TWAI_TIMING_CONFIG_125KBITS();
+
+    // TWAI driver filter configuration, accept all frames.
+    twai_filter_config_t f_config = TWAI_FILTER_CONFIG_ACCEPT_ALL();
+
+    // TWAI driver general configuration.
+    twai_general_config_t g_config =
+        TWAI_GENERAL_CONFIG_DEFAULT(esp_bl_state.tx_pin, esp_bl_state.rx_pin,
+                                    TWAI_MODE_NORMAL);
+    g_config.tx_queue_len = config_can_tx_buffer_size();
+    g_config.rx_queue_len = config_can_rx_buffer_size();
+
+    LOG(BOOTLOADER_LOG_LEVEL, "[Bootloader] Configuring TWAI driver");
+    ESP_ERROR_CHECK(twai_driver_install(&g_config, &t_config, &f_config));
+    LOG(BOOTLOADER_LOG_LEVEL, "[Bootloader] Starting TWAI driver");
+    ESP_ERROR_CHECK(twai_start());
+    esp_bl_state.twai_initialized = true;
+}
+
+/// Callback from the bootloader for entering the application.
+///
+/// This will default to reboot the ESP32.
+void application_entry(void)
+{
+    LOG(BOOTLOADER_LOG_LEVEL, "[Bootloader] application_entry");
+
+    // reset the RTC persistent variable to not enter bootloader on next
+    // restart.
+    bootloader_request = RTC_BOOL_FALSE;
+
+    // restart the esp32 since we do not have a way to rerun app_main.
+    esp_restart();
+}
+
+/// Callback from the bootloader when a reboot should be triggered.
+///
+/// Currently a NO-OP.
+void bootloader_reboot(void)
+{
+    LOG(BOOTLOADER_LOG_LEVEL, "[Bootloader] Reboot requested");
```

Comment:
esp_restart is being done from esp32_bootloader_run, it can be moved to here along with de-init calls but it could reduce some flexibility in case the node needs / wants to do some other tasks prior to reboot after exiting the bootloader code.